worksheet.rb

```diff
@@ -1,34 +1,112 @@
+# Step 3: Make the data structure!
 # Setup the entire data structure:
 # based off of the notes you have above, create the
 # and manually write in data presented in rides.csv
 # You should be copying and pasting the literal data
 # into this data structure, such as "DR0004"
 # and "3rd Feb 2016" and "RD0022"
 
-########################################################
-# Step 4: Total Driver's Earnings and Number of Rides
+# ANSWER: see code below
 
+arr_of_hashes = [
+  { driver_id: "DR0004", date: [02,03,2016], cost: 5, rider_id: "RD0022", rating: 5 },
+  { driver_id: "DR0001", date: [02,03,2016], cost: 10, rider_id: "RD0003", rating: 3 },
+  { driver_id: "DR0002", date: [02,03,2016], cost: 25, rider_id: "RD0073", rating: 5 },
+  { driver_id: "DR0001", date: [02,03,2016], cost: 30, rider_id: "RD0015", rating: 4 },
+  { driver_id: "DR0003", date: [02,04,2016], cost: 5, rider_id: "RD0066", rating: 5 },
+  { driver_id: "DR0004", date: [02,04,2016], cost: 10, rider_id: "RD0022", rating: 4 },
+  { driver_id: "DR0002", date: [02,04,2016], cost: 15, rider_id: "RD0013", rating: 1 },
+  { driver_id: "DR0003", date: [02,05,2016], cost: 50, rider_id: "RD0003", rating: 2 },
+  { driver_id: "DR0002",  date: [02,05,2016], cost: 35, rider_id: "RD0066", rating: 3 },
+  { driver_id: "DR0004", date: [02,05,2016], cost: 20, rider_id: "RD0073", rating: 5 },
+  { driver_id: "DR0001", date: [02,05,2016], cost: 45, rider_id: "RD0003", rating: 2 },
+  ]
+
+########################################################################################################################################################################
+# Step 4: # Step 4: Total Driver's Earnings and Number of Rides
 # Use an iteration blocks to print the following answers:
 # - the number of rides each driver has given
 # - the total amount of money each driver has made
 # - the average rating for each driver
 # - Which driver made the most money?
-# - Which driver has the highest average rating?
\ No newline at end of file
+# - Which driver has the highest average rating?
+
+# Defines i to use as iteration index, and 3 arrays that will hold total number of rides per driver,
+# total earned per driver, total number of rating points per driver, and average rating of each driver, respectively
+i = 0
+total_rides = [0,0,0,0]
+total_earned = [0,0,0,0]
+total_points_rating = [0.0,0.0,0.0,0.0]
+average_rating = [0.0,0.0,0.0,0.0]
+
+# Defines and initiates j with a value of 0 then assigns value for j depending on driver id invoked within the loop
+until i == arr_of_hashes.length do
+  
+  j = 0
+  
+  if arr_of_hashes[i][:driver_id] == "DR0001"
+    j = 0
+  elsif arr_of_hashes[i][:driver_id] == "DR0002"
+    j = 1
+  elsif arr_of_hashes[i][:driver_id] == "DR0003"
+    j = 2
+  elsif arr_of_hashes[i][:driver_id] == "DR0004"
+    j = 3
+  end
```

Comment:
This is clever! consider parsing the driver_id and using the last character - 1 as the value for j. Again, this would allow your code to be a bit more flexible and work even if more drivers were added.

Reply:
I think you could replace this chained conditional with:

```
j = arr_of_hashes[i][:driver_id][-1].to_i
```